File: javascript/two pointers/0042. Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let i=0
    let j=height.length-1
    let maxRight=height[j]
    let maxLeft=height[i]
    let res=0
    while(i<j){
        if(height[i]<=height[j]){
            i++;
            maxLeft=Math.max(maxLeft,height[i])
            res+=maxLeft-height[i]
        }
        else{
            j--;
            maxRight=Math.max(maxRight,height[j])
            res+=maxRight-height[j]
        }
    }
    return res;
};

File: C++/two pointers/0042. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i=0;
        int j=height.size()-1;
        int maxLeft=height[i];
        int maxRight=height[j];
        int res=0;
        while(i<j){
            if(height[i]<height[j])
            {
                i++;
                maxLeft=max(height[i],maxLeft);
                res+=maxLeft-height[i];
            }
            else{
                j--;
                maxRight=max(height[j],maxRight);
                res+=maxRight-height[j];
            }
        }
        return res;
    }
};
//Time complexity:O(N) ,space complexity:O(1)